src/gateway.c

```diff
@@ -288,20 +299,41 @@ static int handle_prepare(struct handle *req)
 		tracef("handle prepare sqlite prepare failed %d", rc);
 		stmt__registry_del(&g->stmts, stmt);
 		failure(req, rc, sqlite3_errmsg(g->leader->conn));
-		return 0;
+		return 0; // ignored
 	}
 
 	if (stmt->stmt == NULL) {
 		tracef("handle prepare empty statement");
 		stmt__registry_del(&g->stmts, stmt);
 		failure(req, rc, "empty statement");
-		return 0;
+		return 0; // ignored
 	}
 
 	response.db_id = (uint32_t)request.db_id;
 	response.id = (uint32_t)stmt->id;
 	response.params = (uint64_t)sqlite3_bind_parameter_count(stmt->stmt);
 	SUCCESS(stmt, STMT);
+	return 0; // ignored
+}
+
+static void prepareBarrierCb(struct barrier *barrier, int status)
+{
+	tracef("prepare barrier cb status:%d", status);
+	(void)prepareBarrierCbInner(barrier, status);
+}
+
+static int handle_prepare(struct handle *req)
+{
+	tracef("handle prepare");
+	struct gateway *g = req->gateway;
+	int rc;
+	CHECK_LEADER(req);
+	g->req = req;
+	rc = leader__barrier(g->leader, &g->barrier, prepareBarrierCb);
+	if (rc != 0) {
+		tracef("handle prepare barrier failed %d", rc);
```

Comment:
I think you need to set `g->req` to `NULL` here. Is this line covered by a test? Would be great to add one if it isn't.

Reply:
Good call -- for testing, how could we inject a failure of `leader__barrier`?

Reply:
There's this `test_heap_fault_enable` e.g. in `test/unit/test_vfs.c`. We could look for a simple mocking framework too.

Reply:
Afaict the stuff in `src/lib/heap.{c,h}` (used by `test_vfs.c`) only deals with the sqlite3 allocator, but it should be straightforward to set up what we need using `raft_heap_set` (already public API).